gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua

```diff
@@ -0,0 +1,150 @@
+  context.skip_apicast_access = true
+  context.skip_apicast_post_action = true
+end
+
+local function report(_, service_id, backend, reports_batcher)
+  local reports = reports_batcher:get_all(service_id)
+
+  -- TODO: verify if we should limit the number of reports sent in a sigle req
+  reporter.report(reports, service_id, backend, reports_batcher)
+end
+
+-- This starts a timer on each worker.
+-- Starting a timer on each worker means that there will be more calls to
+-- 3scale backend, and the config param 'batch_report_seconds' becomes
+-- more confusing because the reporting frequency will be affected by the
+-- number of APIcast workers.
+-- If we started a timer just on one of the workers, it could die, and then,
+-- there would not be any reporting.
+local function ensure_report_timer_on(self, service_id, backend)
+  local check_timer = self.semaphore_report_timer:wait(0)
+
+  if check_timer then
+    if not self.report_timer_on then
+      ngx.timer.every(self.batch_reports_seconds, report,
+        service_id, backend, self.reports_batcher)
+
+      self.report_timer_on = true
+    end
+
+    self.semaphore_report_timer:post()
+  end
+end
+
+local function rejection_reason_from_headers(response_headers)
+  return response_headers and response_headers['3scale-rejection-reason']
+end
+
+local function error(service, rejection_reason)
+  if rejection_reason == 'limits_exceeded' then
+    return errors.limits_exceeded(service)
+  else
+    return errors.authorization_failed(service)
+  end
+end
+
+-- Note: when an entry in the cache expires, there might be several requests
+-- with those credentials and all of them will call auth() on backend with the
+-- same parameters until the auth status is cached again. In the future, we
+-- might want to introduce a mechanism to avoid this and reduce the number of
+-- calls to backend.
+function _M:access(context)
+  local backend = backend_client:new(context.service, http_ng_resty)
+  local usage = context.usage
+  local service = context.service
+  local service_id = service.id
+  local credentials = context.credentials
+
+  ensure_report_timer_on(self, service_id, backend)
+
+  local cached_auth = self.auths_cache:get(service_id, credentials, usage)
+
+  if not cached_auth then
+    local formatted_usage = format_usage(usage)
+    local backend_res = backend:authorize(formatted_usage, credentials)
+    local backend_status = backend_res.status
+
+    if backend_status == 200 then
+      self.auths_cache:set(service_id, credentials, usage, 200)
+      local to_batch = { service_id = service_id, credentials = credentials, usage = usage }
+      self.reports_batcher:add(to_batch.service_id, to_batch.credentials, to_batch.usage)
+    elseif backend_status >= 400 and backend_status < 500 then
+      local rejection_reason = rejection_reason_from_headers(backend_res.headers)
+      self.auths_cache:set(service_id, credentials, usage, backend_status, rejection_reason)
```

Comment:
If `service_id`, `credentials` and `usage` are always passed together to the `auths_cache` then they might deserve own object. And that can be reused to the reports batcher ?

This is clear parameter coupling and makes the parameter list really long.

Reply:
Yes, I thought about that too. I think there are some refactoring opportunities there. Maybe we could extract some classes like `Report` or `Authorization`. I'd rather leave this for a future PR though.